stuck on client.Do when running test

```
Running tests...
panic: test timed out after 3m0s

goroutine 15 [running]:
testing.func008()
	/usr/local/go/src/testing/testing.go:681 +0x168
created by time.goFunc
	/usr/local/go/src/time/sleep.go:129 +0x59

goroutine 1 [chan receive]:
testing.RunTests(0x9064a8, 0x13b2b20, 0x31, 0x31, 0xc208076101)
	/usr/local/go/src/testing/testing.go:556 +0xdc3
testing.(*M).Run(0xc20804e0f0, 0x13bcf00)
	/usr/local/go/src/testing/testing.go:485 +0xe1
main.main()
	github.com/coreos/etcd/client/_test/_testmain.go:200 +0x382

goroutine 12 [chan receive]:
github.com/coreos/etcd/client.(*simpleHTTPClient).Do(0xc208078180, 0x7f7ef8a8b1c0, 0xc20803a580, 0x7f7ef8a8b140, 0x13bd8f0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	github.com/coreos/etcd/client/_test/_obj_test/client.go:291 +0x819
github.com/coreos/etcd/client.TestSimpleHTTPClientDoCancelContextResponseBodyClosed(0xc208060360)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/client/client_test.go:211 +0x762
testing.tRunner(0xc208060360, 0x13b2b68)
	/usr/local/go/src/testing/testing.go:447 +0x134
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:555 +0xd4e

goroutine 13 [chan receive]:
github.com/coreos/etcd/client.func路006()
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/client/client_test.go:205 +0x83
created by github.com/coreos/etcd/client.TestSimpleHTTPClientDoCancelContextResponseBodyClosed
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/client/client_test.go:209 +0x6ed

goroutine 14 [chan receive]:
github.com/coreos/etcd/client.(*fakeTransport).RoundTrip(0xc20800a5c0, 0xc20801eb60, 0x4, 0x0, 0x0)
	/home/travis/gopath/src/github.com/coreos/etcd/gopath/src/github.com/coreos/etcd/client/client_test.go:118 +0x247
github.com/coreos/etcd/client.func路003()
	github.com/coreos/etcd/client/_test/_obj_test/client.go:274 +0xca
created by github.com/coreos/etcd/client.(*simpleHTTPClient).Do
	github.com/coreos/etcd/client/_test/_obj_test/client.go:277 +0x30c
FAIL	github.com/coreos/etcd/client	180.024s
ok  	github.com/coreos/etcd/discovery	0.106s	coverage: 65.7% of statements
ok  	github.com/coreos/etcd/error	0.013s	coverage: 84.6% of statements
ok  	github.com/coreos/etcd/etcdctl/command	0.016s	coverage: 1.2% of statements
ok  	github.com/coreos/etcd/etcdmain	0.027s	coverage: 30.4% of statements
ok  	github.com/coreos/etcd/etcdserver	0.067s	coverage: 59.0% of statements
ok  	github.com/coreos/etcd/etcdserver/etcdhttp	0.038s	coverage: 88.3% of statements
ok  	github.com/coreos/etcd/etcdserver/etcdhttp/httptypes	0.013s	coverage: 78.9% of statements
ok  	github.com/coreos/etcd/migrate	0.017s	coverage: 15.6% of statements
ok  	github.com/coreos/etcd/pkg/fileutil	0.048s	coverage: 80.6% of statements
ok  	github.com/coreos/etcd/pkg/flags	0.014s	coverage: 88.3% of statements
ok  	github.com/coreos/etcd/pkg/idutil	0.011s	coverage: 100.0% of statements
ok  	github.com/coreos/etcd/pkg/ioutil	0.010s	coverage: 100.0% of statements
ok  	github.com/coreos/etcd/pkg/netutil	0.011s	coverage: 90.5% of statements
ok  	github.com/coreos/etcd/pkg/osutil	0.014s	coverage: 87.1% of statements
ok  	github.com/coreos/etcd/pkg/pbutil	0.010s	coverage: 100.0% of statements
ok  	github.com/coreos/etcd/pkg/types	0.014s	coverage: 100.0% of statements
ok  	github.com/coreos/etcd/pkg/transport	0.076s	coverage: 82.4% of statements
ok  	github.com/coreos/etcd/pkg/wait	0.074s	coverage: 100.0% of statements
ok  	github.com/coreos/etcd/proxy	0.015s	coverage: 88.2% of statements
ok  	github.com/coreos/etcd/raft	1.167s	coverage: 82.3% of statements
ok  	github.com/coreos/etcd/snap	0.019s	coverage: 81.7% of statements
ok  	github.com/coreos/etcd/store	0.062s	coverage: 89.0% of statements
ok  	github.com/coreos/etcd/wal	0.210s	coverage: 80.2% of statements
travis_time:end:0963e8f5:start=1425596057075731431,finish=1425596260361845542,duration=203286114111
[0K
```
/cc  @bcwaldon

--- a/client/client_test.go
+++ b/client/client_test.go
@@ -26,6 +26,7 @@ import (
 	"time"
 
 	"github.com/coreos/etcd/Godeps/_workspace/src/golang.org/x/net/context"
+	"github.com/coreos/etcd/pkg/testutil"
 )
 
 type actionAssertingHTTPClient struct {
@@ -113,10 +114,15 @@ func (t *fakeTransport) RoundTrip(*http.Request) (*http.Response, error) {
 	case err := <-t.errchan:
 		return nil, err
 	case <-t.startCancel:
+		select {
+		// this simulates that the request is finished before cancel effects
+		case resp := <-t.respchan:
+			return resp, nil
 		// wait on finishCancel to simulate taking some amount of
 		// time while calling CancelRequest
-		<-t.finishCancel
-		return nil, errors.New("cancelled")
+		case <-t.finishCancel:
+			return nil, errors.New("cancelled")
+		}
 	}
 }
 
@@ -203,12 +209,12 @@ func TestSimpleHTTPClientDoCancelContextResponseBodyClosed(t *testing.T) {
 
 	body := &checkableReadCloser{ReadCloser: ioutil.NopCloser(strings.NewReader("foo"))}
 	go func() {
-		// wait for CancelRequest to be called, informing us that simpleHTTPClient
-		// knows the context is already timed out
-		<-tr.startCancel
+		// wait that simpleHTTPClient knows the context is already timed out,
+		// and calls CancelRequest
+		testutil.WaitSchedule()
 
+		// response is returned before cancel effects
 		tr.respchan <- &http.Response{Body: body}
-		tr.finishCancel <- struct{}{}
 	}()
 
 	_, _, err := c.Do(ctx, &fakeAction{})

